File: lib/previous_matches.rb
# typed: strict
require 'sorbet-runtime'
require 'csv'
require 'set'

# Parses a CSV of previous-year matches into a set of pair-keys, used to stop
# the same mentor/mentee pairing from repeating in a later year's match.
#
# Expects the same columns this program's own match export produces:
# `mentor_email` and `mentee_email` (any other columns, e.g. names, are ignored).
class PreviousMatches
  extend T::Sig

  MentorEmailCol = 'mentor_email'
  MenteeEmailCol = 'mentee_email'

  # Order-independent: a pairing is excluded regardless of which person is the
  # mentor vs. mentee this year.
  sig { params(mentor_email: String, mentee_email: String).returns(String) }
  def self.pair_key(mentor_email, mentee_email)
    [mentor_email.downcase.strip, mentee_email.downcase.strip].sort.join('|')
  end

  sig { params(csv_path: String).returns(T::Set[String]) }
  def self.parse(csv_path)
    rows = T.let(CSV.read(csv_path), T::Array[T::Array[T.nilable(String)]])
    headers = T.must(rows[0]).map { |h| h&.downcase&.strip }

    schema = T.let({}, T::Hash[String, Integer])
    [MentorEmailCol, MenteeEmailCol].each do |col|
      index = headers.index(col)
      if index.nil?
        raise "Previous matches CSV is missing a `#{col}` column. Expected headers: #{MentorEmailCol}, #{MenteeEmailCol}."
      end
      schema[col] = index
    end

    T.must(rows[1..]).reject { |row| row.all?(&:nil?) }.map do |row|
      pair_key(T.must(row[schema.fetch(MentorEmailCol)]), T.must(row[schema.fetch(MenteeEmailCol)]))
    end.to_set
  end
end

File: src/mentor_matching/previous_matches.py
from __future__ import annotations

import csv


class PreviousMatches:
    MENTOR_EMAIL_COL = "mentor_email"
    MENTEE_EMAIL_COL = "mentee_email"

    @classmethod
    def pair_key(cls, mentor_email: str, mentee_email: str) -> str:
        return "|".join(sorted((mentor_email.lower().strip(), mentee_email.lower().strip())))

    @classmethod
    def parse(cls, csv_path: str) -> set[str]:
        with open(csv_path, newline="", encoding="utf-8") as handle:
            rows = list(csv.reader(handle))

        if not rows:
            raise RuntimeError("Previous matches CSV is empty.")

        headers = [header.lower().strip() if header is not None else None for header in rows[0]]
        schema: dict[str, int] = {}
        for column in (cls.MENTOR_EMAIL_COL, cls.MENTEE_EMAIL_COL):
            try:
                schema[column] = headers.index(column)
            except ValueError as exc:
                raise RuntimeError(
                    "Previous matches CSV is missing a "
                    f"`{column}` column. Expected headers: "
                    f"{cls.MENTOR_EMAIL_COL}, {cls.MENTEE_EMAIL_COL}."
                ) from exc

        pair_keys: set[str] = set()
        for row in rows[1:]:
            if not any(value is not None and value != "" for value in row):
                continue
            mentor_email = row[schema[cls.MENTOR_EMAIL_COL]]
            mentee_email = row[schema[cls.MENTEE_EMAIL_COL]]
            pair_keys.add(cls.pair_key(mentor_email, mentee_email))

        return pair_keys
